iidm/iidm-xml-converter/src/main/java/com/powsybl/iidm/xml/GeneratorXml.java

```diff
@@ -63,30 +64,53 @@ protected GeneratorAdder createAdder(VoltageLevel vl) {
 
     @Override
     protected Generator readRootElementAttributes(GeneratorAdder adder, NetworkXmlReaderContext context) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected void readElement(String id, GeneratorAdder adder, NetworkXmlReaderContext context) throws XMLStreamException {
+        Rc rc = readRootElementAttributesGenerator(adder, context);
+        Generator g = adder.add();
+        readPQ(null, g.getTerminal(), context.getReader());
+        readSubElementsGenerator(g, context);
+        IidmXmlUtil.runUntilMaximumVersion(IidmXmlVersion.V_1_5, context, () -> {
+            if (rc.voltageRegulatorOn && g.getRegulatingTerminal() == null) {
+                g.setRegulatingTerminal(g.getTerminal());
+            }
+        });
+        if (rc.voltageRegulatorOn) {
+            g.setVoltageRegulatorOn(true);
+            g.setTargetQ(rc.targetQ);
+        }
+    }
+
+    private static Rc readRootElementAttributesGenerator(GeneratorAdder adder, NetworkXmlReaderContext context) {
         EnergySource energySource = EnergySource.valueOf(context.getReader().getAttributeValue(null, "energySource"));
         double minP = XmlUtil.readDoubleAttribute(context.getReader(), "minP");
         double maxP = XmlUtil.readDoubleAttribute(context.getReader(), "maxP");
         double ratedS = XmlUtil.readOptionalDoubleAttribute(context.getReader(), "ratedS");
-        boolean voltageRegulatorOn = XmlUtil.readBoolAttribute(context.getReader(), "voltageRegulatorOn");
+        Rc rc = new Rc();
+        rc.voltageRegulatorOn = XmlUtil.readBoolAttribute(context.getReader(), "voltageRegulatorOn");
         double targetP = XmlUtil.readDoubleAttribute(context.getReader(), "targetP");
         double targetV = XmlUtil.readOptionalDoubleAttribute(context.getReader(), "targetV");
-        double targetQ = XmlUtil.readOptionalDoubleAttribute(context.getReader(), "targetQ");
+        rc.targetQ = XmlUtil.readOptionalDoubleAttribute(context.getReader(), "targetQ");
+        double targetQ = rc.targetQ;
+        if (Double.isNaN(targetQ)) {
+            targetQ = 0.0;
+        }
         readNodeOrBus(adder, context);
-        Generator g = adder.setEnergySource(energySource)
+        adder.setEnergySource(energySource)
                 .setMinP(minP)
                 .setMaxP(maxP)
                 .setRatedS(ratedS)
-                .setVoltageRegulatorOn(voltageRegulatorOn)
                 .setTargetP(targetP)
                 .setTargetV(targetV)
                 .setTargetQ(targetQ)
-                .add();
-        readPQ(null, g.getTerminal(), context.getReader());
-        return g;
+                .setVoltageRegulatorOn(false);
+        return rc;
     }
 
-    @Override
-    protected void readSubElements(Generator g, NetworkXmlReaderContext context) throws XMLStreamException {
+    private void readSubElementsGenerator(Generator g, NetworkXmlReaderContext context) throws XMLStreamException {
```

Comment:
Same remark

Reply:
Finally, the standard structure of reading has been preserved.